### DISCUSSION TITLE
Allow different EPMD ports

### DISCUSSION THREAD

**Discussion by jonenst** [rabbitmq/rabbitmq-server#12432] [Ideas] [answered]
### Is your feature request related to a problem? Please describe.

I'm always frustrated when I can't freely choose the ports of listening servers. https://www.rabbitmq.com/docs/networking#epmd-port says "All hosts in a cluster must use the same [EPMD] port.". But for various reasons, you may want to use different ports for the different servers:
- on one computer a port is available but on another computer it is not. This can be frustrating if you decide to use a computer later on and its port is not available, you either have to move the port for everyone or find another computer.
- when running on a single computer, you may want to run multiple servers to simulate having multiple servers
- when running on a single computer using docker, you may want to force all traffic to go through the host with mapped ports.

### Describe the solution you'd like

specify the epmd port of each target server:
```
rabbitmqctl join_cluster rabbit@my-rabbit-01.localdomain:4370
```

### Describe alternatives you've considered

when using docker on a single computer, use a docker network for open access between nodes without going through the host

Rejected solutions:
- when running on a single computer, allow the official community rabbitmq docker image from dockerhub to use a single external shared epmd server and provide clear instructions on how to set it up.
- write my own epmd reverse-proxy that can correctly dispatch connections to upstream servers running on different ports based on the name
- run virtual machines
- buy more servers
- use another message broker

**Comment by lukebakken** [comment]
Hello, thanks for using RabbitMQ.

This is the first time I've seen this request come up, so it's very unlikely we will spend time on it unless a customer that pays for a commercial license requests it.

More likely, we would remove the `epmd` requirement from RabbitMQ altogether at some point.

@michaelklishin anything to add?

**Reply by mkuratczyk** [reply]
I was planning on suggesting the same: we basically don't need `epmd` at this point (except, perhaps, for local development). However, I'm not sure whether this solves the problem - we currently use `epmd` to "discover" the port which RabbitMQ uses, which is basically always 25672. Therefore, we don't need `epmd` (we know the answer it'll provide) but perhaps it just moves the problem to a different place - all nodes use 25672.

**Reply by lukebakken** [reply]
Yep. We could allow per-node disterl ports, but then we'd have to have a way to configure the other nodes with that information (peer discovery) and really that's not worth working on unless someone wants to pay for it.

**Reply by jonenst** [reply]
Thank you for your answer. I totally understand that you don't want to add this feature.  I just wanted to share what felt like a valid use case (see my other reply for more details). Thanks again

**Comment by michaelklishin** [comment, marked as answer]
@jonenst you can override the port used by RabbitMQ nodes to receive traffic from peers and CLI tools. You can also override `epmd` port for any Erlang node, including RabbitMQ. 

Both are documented: 

1. [Inter-node Communication Port Range](https://www.rabbitmq.com/docs/networking#epmd-inet-dist-port-range)
2. [`ERL_EPMD_PORT`](https://www.erlang.org/docs/20/man/epmd).

Now that we require a recent version of Erlang, and that supports "epmd-less" clustering, we can simply move to use that in 4.1 or 4.2.

**Reply by jonenst** [reply]
Thank you for taking the time to write a comprehensive answer. I had seen in the docs that it was possible to change the ports, notably the epmd port. But I didn't want one changed epmd port for the cluster, I wanted different changed ports for each node (I didn't even want to change what port nodes were listenening on, only what port they were reachable at)

This would have allowed me to start multiple rabbitmq nodes using docker on a single computer with all traffic going trough the host like this:
```
$ docker run -p 5673:5672 -p 4370:4369 -p 25673:25672 -p 35673:35672 rabbitmq
$ docker run -p 5674:5672 -p 4371:4369 -p 25674:25672 -p 35674:35672  rabbitmq
$ docker run -p 5675:5672 -p 4372:4369 -p 25675:25672 -p 35675:35672  rabbitmq
# and then run join commands using the host mapped ports 
```
Instead I had to create a docker network to allow each container to bind on 4369 (this also had the advantage to not have to list the other ports, but when exploring I precisely wanted to see and know about all the ports).
It was not hard to do so because networks are readily available in docker, so I totally understand if you don't add this feature. I just wanted to share what felt like a valid use case (again mainly exploration and inspection that no other ports are needed). Thanks again

**Reply by michaelklishin** [reply]
Local clusters (outside of Docker) is something that our team uses every single day. [All of the ports used by RabbitMQ](https://www.rabbitmq.com/docs/networking#ports) can be overridden, including the port nodes listen on for peer (node and CLI tool) connections.

If you use containers, it might be easier to leave the ports as is and forward them differently. There's nothing wrong with that.

**Reply by jonenst** [reply]
hi @michaelklishin thanks again for your answer. I had seen the docs you linked to. When you are running local clusters outside of docker on a single computer, are you using a single epmd server for multiple rabbitmq nodes ? In this case you don't have the problem of having to bind multiple times on the the epmd port.

If you use containers, you can't forward them differently because the epmd port must be the same for all empd servers in the cluster (which is what the docs say here https://www.rabbitmq.com/docs/networking#epmd-port or is that wrong ?) and you have one epmd server per container. Could you give an example of using docker on a single computer without using docker networks and having traffic going through differently mapped port on the host ?

If you don't have time that's ok too, I don't want to impose.

Thanks.

**Reply by michaelklishin** [reply]
@jonenst yes, all nodes use the same `epmd` and register different ports with it (the inter-node communication port).

**Reply by jonenst** [reply]
and so when running multiple docker containers on a single computer without creating a docker network (only using mapped ports on the host), what do you do ? A single shared epmd daemon but how do you get all nodes to use it from their respective containers ? Multiple epmd daemons but how do you deal with the fact that they must all bind on the same port on the host ?

**Reply by jonenst** [reply]
also https://www.rabbitmq.com/docs/networking#what-is-epmd-and-how-is-it-used says "epmd runs alongside every RabbitMQ node" and sounds like each node must always have its own epmd (not the case as you described when running multiple rabbitmq nodes on the same computer with a shared epmd) (even though with careful reading one can understand that the epmd may be shared, but the logic of when rabbitmq decides to spawn an epmd daemon is opaque). Maybe rephrase this documentation ? or maybe it's clear enough already. or even maybe this is too niche of use case and not worth talking about in the docs, not sure.

**Reply by michaelklishin** [reply]
@jonenst in 99.9% of production clusters you do not colocate N nodes on a single host, so each node does have its own local `epmd`. If N nodes run on the same host, they will all use the same local `epmd` perfectly fine.

I don't see any need to rephrase the documentation.
